exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java

```diff
@@ -387,62 +443,100 @@ private InfoSnapshot makeTargetInfo(Resource resource) {
+                  null, // resource attributes are only copied for point's attributes
+                  scope,
+                  scope.getAttributes())));
     }
     return new InfoSnapshot(new MetricMetadata("otel_scope"), prometheusScopeInfos);
   }
 
   /**
    * Convert OpenTelemetry attributes to Prometheus labels.
    *
+   * @param resourceAttributes optional resource attributes to be converted.
    * @param scope will be converted to {@code otel_scope_*} labels if {@code otelScopeEnabled} is
    *     {@code true}.
    * @param attributes the attributes to be converted.
    * @param additionalAttributes optional list of key/value pairs, may be empty.
    */
+  @SuppressWarnings({"rawtypes", "unchecked"})
   private Labels convertAttributes(
+      @Nullable Attributes resourceAttributes,
       @Nullable InstrumentationScopeInfo scope,
       Attributes attributes,
       String... additionalAttributes) {
-    int numberOfScopeAttributes = 0;
-    if (otelScopeEnabled && scope != null) {
-      numberOfScopeAttributes = scope.getVersion() == null ? 1 : 2;
-    }
-    String[] names =
-        new String[attributes.size() + numberOfScopeAttributes + additionalAttributes.length / 2];
-    String[] values = new String[names.length];
-    int[] pos = new int[] {0}; // using an array because we want to increment in a forEach() lambda.
+
+    List<AttributeKey> allowedAttributeKeys =
+        addResourceAttributesAsLabels
+            ? filterAllowedResourceAttributeKeys(resourceAttributes)
+            : Collections.emptyList();
+
+    ArrayList<String> names = new ArrayList<>();
+    ArrayList<String> values = new ArrayList<>();
     attributes.forEach(
         (key, value) -> {
-          names[pos[0]] = sanitizeLabelName(key.getKey());
-          values[pos[0]] = value.toString();
-          pos[0]++;
+          names.add(sanitizeLabelName(key.getKey()));
+          values.add(value.toString());
         });
+
     for (int i = 0; i < additionalAttributes.length; i += 2) {
-      names[pos[0]] = additionalAttributes[i];
-      values[pos[0]] = additionalAttributes[i + 1];
-      pos[0]++;
+      names.add(requireNonNull(additionalAttributes[i]));
+      values.add(requireNonNull(additionalAttributes[i + 1]));
     }
+
     if (otelScopeEnabled && scope != null) {
-      names[pos[0]] = OTEL_SCOPE_NAME;
-      values[pos[0]] = scope.getName();
-      pos[0]++;
+      names.add(OTEL_SCOPE_NAME);
+      values.add(scope.getName());
       if (scope.getVersion() != null) {
-        names[pos[0]] = OTEL_SCOPE_VERSION;
-        values[pos[0]] = scope.getVersion();
-        pos[0]++;
+        names.add(OTEL_SCOPE_VERSION);
+        values.add(scope.getVersion());
       }
     }
+
+    if (resourceAttributes != null) {
+      for (AttributeKey attributeKey : allowedAttributeKeys) {
+        // Don't write the same attribute twice.
+        if (names.stream().anyMatch(name -> name.equals(attributeKey.getKey()))) {
```

Comment:
We're caching the `allowedAttributeKeys` to avoid re-evaluating each time whether a resource attribute key should be included, and then for safety check iterate through all existing attributes to confirm there are no conflicts.

Some comments:
- Why only do the safety checks for resource attribute conflicts? The same types of conflict could occur with the hardcoded scope attributes and `additionalAttributes` (i.e. `trace_id`, `span_id`)
- Iterating through `names` for each allowed key seems expensive. This operation is really a set different. I.e. supposing we had `Set<String> allowedAttributeKeys` and `Set<String> names`, we'd want to do something like:
 `allowedAttributeKeys.removeAll(names).forEach(..)`
- It seems unavoidable that we'll have to pay extra in terms of compute and memory to support including resource attributes, but I hope we can design it such that the more common use case of _not_ including resource attributes doesn't incur that expense.

Reply:
To support safety checks for all (which makes sense), I've modified it to use `Map` and usage of `putIfAbsent.` 
The typical use case would have `addResourceAttributesAsLabels` be false, which means the `allowedAttributeKeys` to be an empty list:

```
    List<AttributeKey<?>> allowedAttributeKeys =
        addResourceAttributesAsLabels
            ? filterAllowedResourceAttributeKeys(resource)
            : Collections.emptyList();
```

And that means the following code section would do nothing:

```
    if (resource != null) {
      Attributes resourceAttributes = resource.getAttributes();
      for (AttributeKey attributeKey : allowedAttributeKeys) {
        Object attributeValue = resourceAttributes.get(attributeKey);
        if (attributeValue != null) {
          labelNameToValue.putIfAbsent(
              sanitizeLabelName(attributeKey.getKey()), attributeValue.toString());
        }
      }
    }
```

So I don't think there is any extra compute we need to pay.